File: python/tests/mlir/quantum_struct.py
# ============================================================================ #
# Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

# RUN: PYTHONPATH=../../ pytest -rP  %s | FileCheck %s


import pytest
import cudaq
from dataclasses import dataclass

def test_quantum_struct():
  @dataclass
  class patch:
      q : cudaq.qview
      r : cudaq.qview 

  @cudaq.kernel 
  def entry():
      q = cudaq.qvector(2)
      r = cudaq.qvector(2)
      p = patch(q, r)
      h(p.r[0])

  print(entry)

# CHECK-LABEL:   func.func @__nvqpp__mlirgen__entry()
# CHECK:           %[[VAL_0:.*]] = quake.alloca !quake.veq<2>
# CHECK:           %[[VAL_1:.*]] = quake.alloca !quake.veq<2>
# The struq type is erased in this example.
# CHECK:           %[[VAL_2:.*]] = quake.extract_ref %[[VAL_1]][0] : (!quake.veq<2>) -> !quake.ref
# CHECK:           quake.h %[[VAL_2]] : (!quake.ref) -> ()
# CHECK:           return
# CHECK:         }

File: targettests/execution/quantum_struct.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// clang-format off
// RUN: nvq++ --enable-mlir %s -o %t && %t | FileCheck %s
// clang-format on

#include <cudaq.h>
#include <iostream>

struct PureQuantumStruct {
  cudaq::qview<> view1;
  cudaq::qview<> view2;
};

struct Fehu {
  void operator()(cudaq::qview<> v) __qpu__ { h(v); }
};

struct Ansuz {
  void operator()(cudaq::qview<> v) __qpu__ { x(v); }
};

struct Uruz {
  void operator()(PureQuantumStruct group) __qpu__ {
    Ansuz{}(group.view1);
    Fehu{}(group.view1);
    Fehu{}(group.view2);
    Ansuz{}(group.view2);
  }
};

struct Thurisaz {
  void operator()() __qpu__ {
    cudaq::qvector v1(2);
    cudaq::qvector v2(3);
    PureQuantumStruct pqs{v1, v2};
    Uruz{}(pqs);
    mz(v1);
    mz(v2);
  }
};

int main() {
  auto result = cudaq::sample(Thurisaz{});
  int flags[1 << 5] = {0};
  for (auto &&[b, c] : result) {
    int off = std::stoi(b, nullptr, 2);
    if (off >= (1 << 5) || off < 0) {
      std::cout << "Amazingly incorrect: " << b << '\n';
      return 1;
    }
    flags[off] = 1 + c;
  }
  for (int i = 0; i < (1 << 5); ++i) {
    if (flags[i] == 0) {
      std::cout << "FAILED!\n";
      return 1;
    }
  }
  std::cout << "Wahoo!\n";
  return 0;
}

// CHECK: Wahoo
